[Flow-Aggregator] Implementation of flow aggregator build system and structure

dreamtalen opened the pull request:
This is the first PR of flow aggregator implementation, which only contains the build system and structure of it, including the yaml files, github workflow changes and command line tools. For the main logic of flow aggregator, please take a look at this PR (https://github.com/vmware-tanzu/antrea/pull/1596).

srikartati commented on build/images/flowaggregator/Dockerfile:
```diff
@@ -0,0 +1,5 @@
+
+USER root
+
+COPY bin/flow-aggregator /usr/local/bin/
```
Do we expect the consumer of this dockerfile to compile the flow-aggregator code as a pre-requisite?
Maybe it is better to include the compilation steps for flow-aggregator binary in the dockerfile itself as it is being done for Antrea image.

dreamtalen replied:
Thanks, added.

srikartati commented on build/yamls/flow-aggregator/flow-aggregator.yml:
```diff
@@ -0,0 +1,80 @@
+        command:
+          - flow-aggregator
+        name: flow-aggregator
+        image: dreamtalen/flow-aggregator
```
Are you planning to use Antrea org in dockerhub registry to maintain the image?

dreamtalen replied:
Yes, changed the image name to antrea/flow-aggregator

srikartati commented on cmd/flow-aggregator/flow-aggregator.go:
```diff
@@ -0,0 +1,34 @@
+package main
+
+import (
+	aggregator "github.com/vmware-tanzu/antrea/pkg/flowaggregator"
```
Please use the convention followed in Antrea.
Can you request Antonin to enable the github workflows (golang linters, spell check, unit test etc.) for feature/flow-aggregator branch?

dreamtalen replied:
Thanks, done.

srikartati commented on cmd/flow-aggregator/main.go:
```diff
@@ -0,0 +1,70 @@
+
+	"github.com/spf13/cobra"
+	"github.com/vmware-tanzu/antrea/pkg/log"
+	"github.com/vmware-tanzu/antrea/pkg/version"
```
here too and same everywhere.

srikartati commented on cmd/flow-aggregator/options.go:
```diff
@@ -0,0 +1,129 @@
+		}
+
+		// Convert the string input in net.Addr format
+		hostPortAddr := strSlice[0] + ":" + strSlice[1]
```
Better to take care of flow collector having an IPv6 address like it is done here.
https://github.com/vmware-tanzu/antrea/pull/1541
This PR will be checked in soon.

dreamtalen replied:
Thanks, I'm thinking about taking care it in next PR.

srikartati commented on pkg/flowaggregator/flowaggregator.go:
```diff
@@ -0,0 +1,78 @@
+		case <-stopCh:
+			break
+		case <-exportTicker.C:
+			curMsgs := cp.GetMessages()
```
could you move to go-ipfix v0.3.0 and use the message channel in collector instead?

dreamtalen replied:
Sure, this PR is based on a commit of go-ipfix before intermediate process merged. In the second PR which I'm working on now, I have moved to go-ipfix v0.3.0 and some other comments will be addressed in that PR too.

srikartati replied:
Ok. Will wait for other PR, but please try not to make PRs bulky. It may be hard for code-reviewers to review.

srikartati commented on pkg/flowaggregator/flowaggregator.go:
```diff
@@ -0,0 +1,78 @@
+func (fae *flowAggregator) Run(stopCh <-chan struct{}) {
+	// Go-ipfix TCP collector doesn't work during my last test, fixed as udp temporarily.
+	collectAddr, _ := net.ResolveUDPAddr("udp", "0.0.0.0:4739")
+	cp, _ := collector.InitCollectingProcess(collectAddr, 1024, 0)
```
how is the flow-aggregator collector address fed to the flow exporter at antrea-agent?

srikartati commented on pkg/flowaggregator/flowaggregator.go:
```diff
@@ -0,0 +1,78 @@
+	collectAddr, _ := net.ResolveUDPAddr("udp", "0.0.0.0:4739")
+	cp, _ := collector.InitCollectingProcess(collectAddr, 1024, 0)
+
+	ep, err := exporter.InitExportingProcess(fae.flowCollectorAddr, 1, 0)
```
what if exporting process loose connection to flow collector (provided in configMap) or flow collector is restarted?

antoninbas replied:
+1 - can be in a future PR but we should handle these cases, just like the Agent should be robust to connection loss to the collector or aggregator

I believe this should be done in the go-ipfix library though, just like the gRPC library handles it behind the scene (https://godoc.org/google.golang.org/grpc).

srikartati commented on build/yamls/flow-aggregator/flow-aggregator.yml:
```diff
@@ -0,0 +1,80 @@
+    # Provide the flow collector address as string with format <IP>:<port>[:<proto>], where proto is tcp or udp.
+    # If no L4 transport proto is given, we consider tcp as default.
+    # Defaults to "".
+    flowCollectorAddr: "10.96.28.149:4739:udp"
```
You should use default here and not your testing collector address.

srikartati commented on pkg/flowaggregator/flowaggregator.go:
```diff
@@ -0,0 +1,78 @@
+					setSent.AddRecord(record.GetInfoElements(), record.GetTemplateID())
+				}
+				bytesSent, err := ep.AddSetAndSendMsg(setType, setSent)
+				klog.Infof("Msg index: %d, sent bytes: %d, error: %v\n", i, bytesSent, err)
```
Increase the log level 4 for debug messages and use error log if err is not nil.

srikartati reviewed (commented):
Thanks for the PR @dreamtalen 
We should probably change the name of featureGate in Antrea agent as FlowVisbilty instead of FlowExporter to incorporate the flow-aggregator implementation in the same feature gate. What do you think?

srikartati reviewed (commented):
I see the version of go-ipfix is v0.2.1. You should at least move to v0.2.4 which is on master. May be rebasing the feature/flow-aggregator with master helps.

antoninbas commented on build/images/flowaggregator/Dockerfile:
```diff
@@ -0,0 +1,13 @@
+
+RUN make flow-aggregator
+
+FROM antrea/base-ubuntu:2.14.0
```
add metadata (see other Dockerfiles in the repo for examples)

antoninbas commented on build/images/flowaggregator/Dockerfile:
```diff
@@ -0,0 +1,13 @@
+FROM golang:1.15 as flowaggregator-build
```
maybe the directory should be called `flow-aggregator` to match the image name

antoninbas commented on Makefile:
```diff
@@ -274,6 +279,12 @@ octant-antrea-ubuntu:
 	docker tag antrea/octant-antrea-ubuntu:$(DOCKER_IMG_VERSION) antrea/octant-antrea-ubuntu
 
+.PHONY: flow-aggregator-ubuntu
+flow-aggregator-ubuntu:
```
rebase master and match what we did for other Docker images, e.g.:

```
# Build bins in a golang container, and build the antrea-ubuntu Docker image.
.PHONY: build-ubuntu
build-ubuntu:
	@echo "===> Building Antrea bins and antrea/antrea-ubuntu Docker image <==="
ifneq ($(DOCKER_REGISTRY),"")
	docker build -t antrea/antrea-ubuntu:$(DOCKER_IMG_VERSION) -f build/images/Dockerfile.build.ubuntu .
else
	docker build --pull -t antrea/antrea-ubuntu:$(DOCKER_IMG_VERSION) -f build/images/Dockerfile.build.ubuntu .
endif
	docker tag antrea/antrea-ubuntu:$(DOCKER_IMG_VERSION) antrea/antrea-ubuntu
```

cc @lzhecheng

antoninbas commented on .github/workflows/build.yml:
```diff
@@ -96,3 +98,20 @@ jobs:
         make push
         make push-release
+
+  build-flow-aggregator:
```
update `build_tag.yml` as well so that we push tagged images for releases

antoninbas commented on build/yamls/flow-aggregator/flow-aggregator.yml:
```diff
@@ -0,0 +1,80 @@
+apiVersion: v1
```
This manifest needs to be integrated with `hack/generate-manifest.sh`.

In particular `generate-manifest.sh` should support generating a release version of the manifest for which the docker image tag is the correct release tag. This manifest should be automatically uploaded to release assets when a Github release is created (so you need to edit `upload_release_assets.yml`).

Make sure that you double check everything and that you test `hack/generate-manifest.sh` locally, otherwise it breaks the release flow.

antoninbas commented on build/yamls/flow-aggregator/flow-aggregator.yml:
```diff
@@ -0,0 +1,80 @@
+      - name: flow-aggregator-config
+        configMap:
+          name: flow-aggregator-configmap
+        
```
there is an extra empty line here

antoninbas commented:
The manifest needs to be integrated with `hack/generate-manifest` & kustomize. We need to have a dev version of the manifest and a release version of the manifest.

Changes to `upload_release_assets.yml` and `build_tag.yml` can be postponed to a future PR. It's important to get them right since these workflows are not exercised until we make an actual release.

dreamtalen commented:
> The manifest needs to be integrated with `hack/generate-manifest` & kustomize. We need to have a dev version of the manifest and a release version of the manifest.
> 
> Changes to `upload_release_assets.yml` and `build_tag.yml` can be postponed to a future PR. It's important to get them right since these workflows are not exercised until we make an actual release.

Thanks, just addressed these comments.

dreamtalen commented on cmd/flow-aggregator/config.go:
```diff
@@ -0,0 +1,27 @@
+	// records to the flow collector.
+	// Flow export frequency should be greater than or equal to 1s(one second).
+	// Defaults to "60s". Follow the time units of duration.
+	FlowExportFrequency string `yaml:"flowExportFrequency,omitempty"`
```
Thanks, addressed.

lzhecheng commented on build/yamls/flow-aggregator/base/kustomization.yml:
```diff
@@ -0,0 +1,11 @@
+  app: flow-aggregator
+namespace: flow-aggregator
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
```
Better to have an extra empty line here.

dreamtalen replied:
Thanks, done.

lzhecheng commented on build/yamls/flow-aggregator/base/conf/flow-aggregator.conf:
```diff
@@ -0,0 +1,11 @@
+# Flow export interval should be greater than or equal to 1s (one second).
+# Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+# Defaults to "60s".
+flowExportInterval: 60s
```
ditto.

dreamtalen commented on .github/workflows/golicense.yml:
```diff
@@ -4,10 +4,12 @@ on:
     branches:
     - master
     - release-*
+    - feature/flow-aggregator
```
Thanks, remove feature/flow-aggregator in go license and kind update workflow since they are not essential.

srikartati commented on build/yamls/flow-aggregator.yml:
```diff
@@ -0,0 +1,80 @@
+apiVersion: v1
```
What is difference between this file and build/yamls/flow-aggregator/flow-aggregator.yml

dreamtalen replied:
Thanks, build/yamls/flow-aggregator/flow-aggregator.yml has been deleted since we use kustomize to generate build/yamls/flow-aggregator.

srikartati replied:
As there is separate folder for flow-aggregator, why don't we generate at build/yamls/flow-aggregator/flow-aggregator.yml and not at build/yamls/flow-aggregator.yml?

dreamtalen replied:
Yeah this location doesn't matter too much I think, I was following the style in build/yamls/octant/ or windows/, the yaml files they generate are under build/yamls/ but not their own subdirectory.

srikartati replied:
Following others' convention sounds fine to me.

srikartati reviewed (commented):
Please look at the workflow failures corresponding to linters, compiling the binary and unit tests.

antoninbas commented on .github/workflows/build.yml:
```diff
@@ -96,3 +98,20 @@ jobs:
+        DOCKER_PASSWORD: ${{ secrets.DOCKER_PASSWORD }}
+      run: |
+        echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin
+        docker push antrea/flow-aggregator:latest
```
It seems that the PR description is not rally accurate:

> Currently I’m pushing the docker image of flow aggregator to my own docker account for development use. Will push the image into antrea dockerhub after the flow aggregator implementation finished.

antoninbas commented on build/images/flow-aggregator/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+FROM golang:1.15 as flow-aggregator-build
+
+LABEL maintainer="Antrea <projectantrea-dev@googlegroups.com>"
+LABEL description="The docker image for flow aggregator"
```
these label statements should come after the *last* `FROM` statement if I recall correctly?

antoninbas commented on build/yamls/flow-aggregator/patches/release/.gitignore:
```diff
@@ -0,0 +1 @@
+# placeholder
```
I believe the convention is an empty file called `.gitkeep`

dreamtalen replied:
ok, I was following the style in build/yamls/octant/ or windows/, could you please check again?

antoninbas replied:
I added the .gitignore file before I became aware of the .gitkeep "convention", but that's a pretty minor point anyway. Feel free to update all of them to .gitkeep as part of this PR or keep is as it is.

antoninbas commented on cmd/flow-aggregator/config.go:
```diff
@@ -0,0 +1,28 @@
+	// If no L4 transport proto is given, we consider tcp as default.
+	// Defaults to "".
+	FlowCollectorAddr string `yaml:"flowCollectorAddr,omitempty"`
+	// Provide flow export interval as a duration string. This determines how often the flow aggregator exports flow
+	// records to the flow collector.
+	// Flow export interval should be greater than or equal to 1s (one second).
+	// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+	// Defaults to "60s".
+	FlowExportInterval string `yaml:"flowExportInterval,omitempty"`
```
Probably a bit out of scope of this PR, but IIRC Rob Cowart recommended that we spread the export over the time interval (both in the agent and in the aggregator), to avoid all records arriving at the same time

srikartati replied:
Yes using different export intervals for active and inactive flows need a proposal. We need to add the notion of inactive flows in the exported IPFIX fields.

antoninbas replied:
I think this becomes especially important at the aggregator, which will generate more traffic than each individual agent presumably

antoninbas commented on hack/generate-manifest-flow-aggregator.sh:
```diff
@@ -0,0 +1,132 @@
+}
+
+_usage="Usage: $0 [--mode (dev|release)] [--keep] [--help|-h]
+Generate a YAML manifest to run Antrea on Windows Nodes, using Kustomize, and print it to stdout.
```
Update the comment

antoninbas commented on pkg/flowaggregator/flowaggregator.go:
```diff
@@ -0,0 +1,79 @@
+	defer exportTicker.Stop()
+	sentMsgNum := 0
+	go cp.Start()
+	for {
```
so all the logic necessary to store connection state and reconcile flows from source and destination Nodes will go here in a future PR?

antoninbas replied:
you may want to review whether all the log messages in this loop are necessary or have the correct verbosity?

antoninbas commented on pkg/flowaggregator/flowaggregator.go:
```diff
@@ -0,0 +1,79 @@
+func (fae *flowAggregator) Run(stopCh <-chan struct{}) {
+	// Go-ipfix TCP collector doesn't work during my last test, fixed as udp temporarily.
+	collectAddr, _ := net.ResolveUDPAddr("udp", "0.0.0.0:4739")
+	cp, _ := collector.InitCollectingProcess(collectAddr, 1024, 0)
```
why is the error ignored?

antoninbas commented on pkg/flowaggregator/flowaggregator.go:
```diff
@@ -0,0 +1,79 @@
+	collectAddr, _ := net.ResolveUDPAddr("udp", "0.0.0.0:4739")
+	cp, _ := collector.InitCollectingProcess(collectAddr, 1024, 0)
+
+	ep, err := exporter.InitExportingProcess(fae.flowCollectorAddr, 1, 0)
```
is there some symmetry in the go-ipfix implementation for the agent -> aggregator connection and the aggregator -> collector connection? @srikartati

antoninbas reviewed (commented):
some comments, but I'm satisfied with the Github workflow changes and build system changes

dreamtalen commented:
> some comments, but I'm satisfied with the Github workflow changes and build system changes

Antonin, thanks for your comments. The logic in pkg/flowaggregator/flowaggregator.go has changed a lot since the go-ipfix version moved from v0.2.1 -> v0.3.1, so please check the second PR (https://github.com/vmware-tanzu/antrea/pull/1596) about the main logic of flow aggregator. Other comments will be addressed soon.

srikartati reviewed (commented):
@antoninbas Yes there is dependency with go-ipfix versions. That is why there are some errors in workflows too.

@dreamtalen Actually, could you remove this file from the PR entirely and check-in as part of PR #1596 ?

dreamtalen commented:
> @antoninbas Yes there is dependency with go-ipfix versions. That is why there are some errors in workflows too.
> 
> @dreamtalen Actually, could you remove this file from the PR entirely and check-in as part of PR #1596? Or keep some minimal code.

Thanks, I removed most logic in pkg/flowaggregator/flowaggregator.go and will update the description of this commit.

srikartati commented on Makefile:
```diff
@@ -306,6 +312,16 @@ octant-antrea-ubuntu:
 	docker build --pull -t antrea/octant-antrea-ubuntu:$(DOCKER_IMG_VERSION) -f build/images/Dockerfile.octant.ubuntu .
 	docker tag antrea/octant-antrea-ubuntu:$(DOCKER_IMG_VERSION) antrea/octant-antrea-ubuntu
 
+.PHONY: flow-aggregator-ubuntu
```
should we add this target along with `build-ubuntu` to get executed when we do `make`? Or expect folks to build image by using the target directly? @antoninbas what do you think?

srikartati reviewed (commented):
LGTM except for one comment on building the flow-aggregator image through `make`.

srikartati commented:
/test-all

antoninbas commented on build/yamls/flow-aggregator/base/flow-aggregator.yml:
```diff
@@ -0,0 +1,52 @@
+spec:
+  selector:
+    app: flow-aggregator
+  type: NodePort
```
maybe I am missing something, but why is this a `NodePort` Service? a ClusterIP is enough for the agent to connect to the aggregator

dreamtalen replied:
Yes, I have removed the NodePort type declaration in the second PR, sorry for not merging it into this PR.

antoninbas commented on build/yamls/flow-aggregator/base/conf/flow-aggregator.conf:
```diff
@@ -0,0 +1,11 @@
+# Provide the flow collector address as string with format <IP>:<port>[:<proto>], where proto is tcp or udp.
+# If no L4 transport proto is given, we consider tcp as default.
+# Defaults to "".
```
nit: don't think this line is needed

dreamtalen replied:
Sorry, do you mean these description lines are not needed for parameters in flow-aggregator.conf?

antoninbas replied:
yes, I don't believe we typically include them in the .conf file, since the default value is obvious from the commented-out line

dreamtalen replied:
Got it, thanks.

antoninbas commented on build/yamls/flow-aggregator/base/conf/flow-aggregator.conf:
```diff
@@ -0,0 +1,11 @@
+# records to the flow collector.
+# Flow export interval should be greater than or equal to 1s (one second).
+# Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+# Defaults to "60s".
```
ditto

antoninbas commented on pkg/flowaggregator/flowaggregator.go:
```diff
@@ -0,0 +1,39 @@
+	}
+}
+
+func (fae *flowAggregator) Run(stopCh <-chan struct{}) {
```
nit: why `fae` instead of `fa`?

dreamtalen replied:
Changed to `fa`, will also update the variable name in subsequent PRs.

antoninbas commented on cmd/flow-aggregator/options.go:
```diff
@@ -0,0 +1,129 @@
+	}
+	if o.config.FlowCollectorAddr == "" {
+		return fmt.Errorf("IPFIX flow collector address should be provided")
+	} else {
```
if the address parsing logic can be unified with the agent, please consider doing it in a subsequent PR

antoninbas reviewed (commented):
some small comments, otherwise LGTM

main comment on my side is that I don't think the flow aggregator needs to be a NodePort Service?

srikartati reviewed (approved):
LGTM. 
Seems like all comments are addressed. Requesting re-review of @antoninbas

antoninbas reviewed (approved):
LGTM
